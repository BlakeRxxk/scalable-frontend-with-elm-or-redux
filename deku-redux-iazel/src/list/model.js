import { List, Map } from 'immutable'

const init = () => Map({
  items: List(),
  inpts: Map()
})

const get = (state, i) => state.getIn(['items', i])
const set = (state, i, item) => state.setIn(['items', i], item)
const push = (state, item) => state.set('items', state.get('items').push(item))
const getItems = (state) => state.get('items')

const remove = (state, i) => state.set('items', state.get('items').delete(i))

const getNew = (state) => state.get('inpts')
const setNewVal = (state, field, value) => state.setIn(['inpts', field], value)

export {
  get,
  set,
  init,
  push,
  remove,
  getNew,
  setNewVal,
  getItems
}
