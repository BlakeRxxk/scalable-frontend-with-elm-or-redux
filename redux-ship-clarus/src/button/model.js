// @flow

export type State = {
  status: 'green' | 'red',
};

export const initialState: State = {
  status: 'green',
};

export type Patch = {
  type: 'Toggle',
};

export function reduce(state: State, patch: Patch): State {
  switch (patch.type) {
  case 'Toggle':
    return {
      ...state,
      status: state.status === 'green' ? 'red' : 'green',
    };
  default:
    return state;
  }
}
