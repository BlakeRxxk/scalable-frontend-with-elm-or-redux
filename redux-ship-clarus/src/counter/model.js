// @flow

export type State = {
  count: number,
};

export const initialState: State = {
  count: 0,
};

export type Action = {
  type: 'IncrementByOne',
} | {
  type: 'IncrementByTwo',
};

export function reduce(state: State, action: Action): State {
  switch (action.type) {
  case 'IncrementByOne':
    return {
      ...state,
      count: state.count + 1,
    };
  case 'IncrementByTwo':
    return {
      ...state,
      count: state.count + 2,
    };
  default:
    return state;
  }
}
