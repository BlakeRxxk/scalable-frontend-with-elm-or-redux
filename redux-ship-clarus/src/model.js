// @flow
import * as ButtonModel from './button/model';
import * as CounterModel from './counter/model';
import * as RandomGifModel from './random-gif/model';
import * as RandomGifPairModel from './random-gif-pair/model';

export type State = {
  button: ButtonModel.State,
  counter: CounterModel.State,
  randomGif: RandomGifModel.State,
  randomGifPair: RandomGifPairModel.State,
};

export const initialState: State = {
  button: ButtonModel.initialState,
  counter: CounterModel.initialState,
  randomGif: RandomGifModel.initialState,
  randomGifPair: RandomGifPairModel.initialState,
};

export type Patch = {
  button?: ButtonModel.Patch,
  counter?: CounterModel.Patch,
  randomGif?: RandomGifModel.Patch,
  randomGifPair?: RandomGifPairModel.Patch,
};

export function reduce(state: State, patch: Patch): State {
  return {
    ...state,
    ...patch.button &&
      {button: ButtonModel.reduce(state.button, patch.button)},
    ...patch.counter &&
      {counter: CounterModel.reduce(state.counter, patch.counter)},
    ...patch.randomGif &&
      {randomGif: RandomGifModel.reduce(state.randomGif, patch.randomGif)},
    ...patch.randomGifPair &&
      {randomGifPair: RandomGifPairModel.reduce(state.randomGifPair, patch.randomGifPair)},
  };
}
