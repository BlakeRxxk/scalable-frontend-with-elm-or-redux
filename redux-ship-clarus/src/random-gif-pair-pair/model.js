// @flow
import * as RandomGifPairModel from '../random-gif-pair/model';

export type State = {
  first: RandomGifPairModel.State,
  second: RandomGifPairModel.State,
};

export const initialState: State = {
  first: RandomGifPairModel.initialState,
  second: RandomGifPairModel.initialState,
};

export type Patch = {
  first?: RandomGifPairModel.Patch,
  second?: RandomGifPairModel.Patch,
};

export function reduce(state: State, patch: Patch): State {
  return {
    ...state,
    ...patch.first && {first: RandomGifPairModel.reduce(state.first, patch.first)},
    ...patch.second && {second: RandomGifPairModel.reduce(state.second, patch.second)},
  };
}
