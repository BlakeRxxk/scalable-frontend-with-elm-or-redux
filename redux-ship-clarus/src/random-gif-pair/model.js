// @flow
import * as RandomGifModel from '../random-gif/model';

export type State = {
  first: RandomGifModel.State,
  second: RandomGifModel.State,
};

export const initialState: State = {
  first: RandomGifModel.initialState,
  second: RandomGifModel.initialState,
};

export type Patch = {
  first?: RandomGifModel.Patch,
  second?: RandomGifModel.Patch,
};

export function reduce(state: State, patch: Patch): State {
  return {
    ...state,
    ...patch.first && {first: RandomGifModel.reduce(state.first, patch.first)},
    ...patch.second && {second: RandomGifModel.reduce(state.second, patch.second)},
  };
}
