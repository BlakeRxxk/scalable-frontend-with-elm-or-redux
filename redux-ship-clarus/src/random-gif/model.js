// @flow

export type State = {
  gifUrl: ?string,
  isLoading: bool,
};

export const initialState: State = {
  gifUrl: null,
  isLoading: false,
};

export type Action = {
  type: 'LoadStart',
} | {
  type: 'LoadSuccess',
  gifUrl: string,
};

export function reduce(state: State, action: Action): State {
  switch (action.type) {
  case 'LoadStart':
    return {
      ...state,
      isLoading: true,
    };
  case 'LoadSuccess':
    return {
      ...state,
      isLoading: false,
      gifUrl: action.gifUrl,
    };
  default:
    return state;
  }
}
