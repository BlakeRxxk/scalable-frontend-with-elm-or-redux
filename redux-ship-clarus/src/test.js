// @flow
import React from 'react';
import renderer from 'react-test-renderer';

export function dispatch() {
}

export function snapshotApplyCommit<Commit, Patch, State>(
  applyCommit: (state: State, commit: Commit) => Patch,
  configs: {[id: string]: {commit: Commit, state: State}}
): void {
  Object.keys(configs).forEach(id => {
    it(id, () => {
      const {commit, state} = configs[id];
      expect(applyCommit(state, commit)).toMatchSnapshot();
    });
  });
}

export function snapshotComponent<Config>(
  component: ReactClass<Config>,
  configs: {[id: string]: Config}
): void {
  Object.keys(configs).forEach(id => {
    it(id, () => {
      const element = React.createElement(component, configs[id]);
      const tree = renderer.create(element).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
}

export function snapshotReduce<Patch, State>(
  reduce: (state: State, patch: Patch) => State,
  configs: {[id: string]: {patch: Patch, state: State}}
): void {
  Object.keys(configs).forEach(id => {
    it(id, () => {
      const {patch, state} = configs[id];
      expect(reduce(state, patch)).toMatchSnapshot();
    });
  });
}
