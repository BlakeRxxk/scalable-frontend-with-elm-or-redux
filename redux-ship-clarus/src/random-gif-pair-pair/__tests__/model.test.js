// @flow
import * as Test from '../../test';
import * as RandomGifPairPairModel from '../model';

Test.snapshotReduce(RandomGifPairPairModel.reduce, {
  firstSecond: {
    patch: {
      first: {second: {type: 'LoadStart'}},
    },
    state: RandomGifPairPairModel.initialState,
  },
  secondFirst: {
    patch: {
      second: {first: {type: 'LoadStart'}},
    },
    state: RandomGifPairPairModel.initialState,
  },
});
