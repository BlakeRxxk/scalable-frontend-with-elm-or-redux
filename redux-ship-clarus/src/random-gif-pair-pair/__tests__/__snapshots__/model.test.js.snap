exports[`test firstSecond 1`] = `
Object {
  "first": Object {
    "first": Object {
      "gifUrl": null,
      "isLoading": false
    },
    "second": Object {
      "gifUrl": null,
      "isLoading": true
    }
  },
  "second": Object {
    "first": Object {
      "gifUrl": null,
      "isLoading": false
    },
    "second": Object {
      "gifUrl": null,
      "isLoading": false
    }
  }
}
`;

exports[`test secondFirst 1`] = `
Object {
  "first": Object {
    "first": Object {
      "gifUrl": null,
      "isLoading": false
    },
    "second": Object {
      "gifUrl": null,
      "isLoading": false
    }
  },
  "second": Object {
    "first": Object {
      "gifUrl": null,
      "isLoading": true
    },
    "second": Object {
      "gifUrl": null,
      "isLoading": false
    }
  }
}
`;
